text/2021-07-04-selector-plugin.md

```diff
@@ -0,0 +1,144 @@
+
+3. Implementation: The specific chaos implementation, which has been extracted
+   from the controller routine.
+
+However, the first two parts are hard coupling now. We will firstly define their interface, and then describe how could we develop a new selector plugin.
+
+## Routine
+
+When the controller receives a request, it will do following things:
+
+1. Call related selectors to list all selected resources.
+2. Call scope limitation to randomly choose some of them.
+3. Iterate through the selected resources and call the implementation.
+
+## Interface
+
+### Scope limitation
+
+The scope limitation is quite simple now. It accepts a list and return a list.
+As all records Ids are strings, the only method is `Limit(input []string)
+[]string`. This interface could be unstable as we will not design the "scope limitation plugin" in the near future, so that we could add more functions to it when needed.
+
+For example, one day if we need to dynamically add new records, we only need to
+add arguments or methods to this interface.
+
+### Selector
+
+One selector should implement this interface:
+
+```go
+type RawSelector map[string]interface{}
+
+type Selector interface {
+    Validate(selector RawSelector) field.ErrorList
+    Default(root map[string]interface{}, selector RawSelector) RawSelector
+
+    Select(selector RawSelector) ([]string, error)
+}
+```
+
+The controller will pass the raw definition of a selector to the selector
+plugin.
+
+#### Definition
+
+At least, they should be declared in the kubernetes object, so we need to
+integrate a fully customizable component in the existing resources. The
+`unstructured.Unstructured` could be used to represent this situation.
+
+However, it cannot be the embedded fields. For most situation, it is fine as the
+`PodSelector` has a `Selector` field, which could be modified to the
+`unstructured` type. However, for the `AWSSelector` and `GCPSelector`, we have
+to break the compatibility.
+
+It's not a wise choice to bring an unstructured type in the definition, and the
+`unstructured.Unstructured` was designed to serve for the full unstructured
+type, but not partially unstructured object. But we have no choice.
+
+#### Communication
+
+`grpc` is one possible choice. They could communicate through tcp or unix
+socket. The plugin user may need to submit a `Selector` custom resource to
+register the selector plugin:
+
+```yaml
+apiVersion: chaos-mesh.org/v1alpha1
+kind: Selector
+metadata:
+  name: pod-selector
+spec:
+  serviceName: xxxx
+  ...
+```
+
+And then, the controller could establish a `grpc` connection to the service and
+hold a selector client.
+
+Other RPC like jsonrpc or http could also be considered, but I prefer grpc as it
+has a richer ecosystem in golang's world.
+
+Besides the network communication, chaos-mesh should also provide a way to
+communicate through local unix socket, to avoid uneccessary network overhead.
+For example, the plugin can be registered with a path:
```

Comment:
What does this for? I think it is impossible for an external plugin to using unix domain socket: I imagine that the external plugin runs as one pod that you do not know which node will be scheduled on, if the pod of "external plugin" and the pod "controller-manager" run on different node, there is no way to use unix domain socket.

Reply:
Yes. I plan to setup some "official plugin" like pod selector to use unix socket. They use the same SDK with others (but run in the same binary or container or pod with the controller).

 The users can also add extra container and volume to the controller, for performance.